pkg/virt-launcher/metadata/safedata.go

```diff
@@ -23,34 +23,67 @@ import (
 	"sync"
 )
 
-type SafeData[T any] struct {
+type SafeData[T comparable] struct {
 	m           sync.Mutex
 	initialized bool
+	dirtyChanel chan<- struct{}
 	data        T
 }
 
+// Load reads and returns safely the data and a flag.
+// The flag specifies if the data is already initialized (true) or not (false).
+// Data which is not yet initialized has never been stored.
 func (d *SafeData[T]) Load() (T, bool) {
 	d.m.Lock()
 	defer d.m.Unlock()
 	return d.data, d.initialized
 }
 
+// Store persists safely the inputted data.
+// As a side effect, it marks the data as initialized and
+// in case a notification channel exists, a signal is sent.
 func (d *SafeData[T]) Store(data T) {
+	d.Set(data)
+	d.notify()
+}
+
+// Set persists safely the inputted data.
+// As a side effect, it marks the data as initialized.
+//
+// Note: Unlike Store, this method does not have a notification side effect.
+func (d *SafeData[T]) Set(data T) {
```

Comment:
It seems to me this could be characteristic of `SafeData`... What I mean is we can just initialize the `SafeData` either with an "empty" channel or not. WDYT?

Reply:
Not sure I understand.
You can now initialize the safeData with an empty channel, and the notify will be a noop.
However, if you want a specific data to sometime notify and sometime, not, you need a special method.

Reply:
>However, if you want a specific data to sometime notify and sometime, not, you need a special method.

I don't see usage for now (maybe I am missing it). From what I see `UID` don't need the notification. 

Another observation is that we probably need only `Initialize` rather than `Set`. We are using the `Set` only to initialize the cache.

Reply:
The `UID` is indeed a special case, on which we could instantiate without a channel, however, `GracePeriod` is used on both scenarios.

As we need it anyway, I would leave it like this, so it will be clear (in a more direct manner) that a notification is not sent.

> Another observation is that we probably need only Initialize rather than Set. We are using the Set only to initialize the cache.

I think this comes with an assumption that the cache is not read before it is initialized. Or I just do not see the whole solution you are proposing.
The initialized flag specifies at the moment if it was touched already, both store and set touch it.
Also, the store operation, even if it comes and touches it for the first time (initializes it), still needs to send a notification. That is, because most stores occur after the domain is created and operational. So we cannot decide not to notify if it is the first time the data is persisted in the cache.